File: apps/dashboard/models/device.model.ts
import { z } from "zod";
import { CentreModelDataSchema } from "./centre.model";
import { DeviceActivityType, StatusEnum } from "./enums";
import { userModelDataSchema } from "./user.model";
const DeviceActivityModelDataSchema = z.object({
  id: z.string().optional(),
  deviceId: z.string(),
  deviceActivityType: z.nativeEnum(DeviceActivityType),
  centreId: z.string().optional().nullable(),
  actionBy: z.string().optional().nullable(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
  actionByUser: userModelDataSchema.optional().nullable(),
  centre: CentreModelDataSchema.optional().nullable(),
});
export const DeviceModelDataSchema = z.object({
  id: z.string().optional(),
  code: z.string().optional().nullable(),
  codeSequence: z.number().optional().nullable(),
  otoscopeID: z.string().optional().nullable(),
  tabletID: z.string().optional().nullable(),
  deviceID: z.string().optional().nullable(),
  tabletAppVersion: z.string().optional().nullable(),
  tabletAndroidVersion: z.string().optional().nullable(),
  status: z.nativeEnum(StatusEnum),
  pendingUpdate: z.boolean().optional().nullable(),
  pendingLookup: z.boolean().optional().nullable(),
  lastUpdateChecked: z.string().optional().nullable(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
  centreId: z.string().optional().nullable(),
  deviceActivities: z
    .array(DeviceActivityModelDataSchema)
    .optional()
    .nullable(),
  centre: CentreModelDataSchema.optional().nullable(),
});

export const CreateDeviceModelSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  data: DeviceModelDataSchema.nullable().optional(),
});

export const DeviceModelSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  data: z.array(DeviceModelDataSchema),
});

export type DeviceModel = z.infer<typeof DeviceModelSchema>;
export type CreateDeviceModel = z.infer<typeof CreateDeviceModelSchema>;
export type DeviceModelData = z.infer<typeof DeviceModelDataSchema>;

File: apps/earkart_omni/lib/models/device/device.model.dart
import 'dart:convert';

import 'package:earkart_omni/models/centre/centre.entity.dart';
import 'package:earkart_omni/models/device/device.entity.dart';
import 'package:earkart_omni/models/enums.dart';

DeviceModel deviceModelFromJson(String str) =>
    DeviceModel.fromJson(json.decode(str));
String deviceModelToJson(DeviceModel data) => json.encode(data.toJson());

class DeviceModel {
  final bool success;
  final String message;
  final DeviceData? data;

  DeviceModel({required this.success, required this.message, this.data});

  factory DeviceModel.fromJson(Map<String, dynamic> json) {
    return DeviceModel(
      success: json['success'],
      message: json['message'],
      data: json['data'] != null ? DeviceData.fromJson(json['data']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {'success': success, 'message': message, 'data': data?.toJson()};
  }
}

class DeviceData extends DeviceEntity {
  @override
  final String id;
  @override
  final String? code;
  @override
  final int? codeSequence;
  @override
  final String? tabletID;
  @override
  final String? deviceID;
  @override
  final String? otoscopeID;
  @override
  final String? tabletAppVersion;
  @override
  final String? tabletAndroidVersion;
  @override
  final String? centreId;
  @override
  final Status status;
  @override
  final DateTime createdAt;
  @override
  final DateTime updatedAt;
  @override
  final CentreEntity? centre;

  const DeviceData({
    required this.id,
    this.code,
    this.codeSequence,
    this.tabletID,
    this.deviceID,
    this.otoscopeID,
    this.tabletAppVersion,
    this.tabletAndroidVersion,
    this.centreId,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.centre,
  }) : super(
         id: id,
         code: code,
         codeSequence: codeSequence,
         tabletID: tabletID,
         deviceID: deviceID,
         otoscopeID: otoscopeID,
         tabletAppVersion: tabletAppVersion,
         tabletAndroidVersion: tabletAndroidVersion,
         centreId: centreId,
         status: status,
         createdAt: createdAt,
         updatedAt: updatedAt,
         centre: centre,
       );

  factory DeviceData.fromJson(Map<String, dynamic> json) {
    return DeviceData(
      id: json['id'] ?? '',
      code: json['code'],
      codeSequence: json['codeSequence'],
      tabletID: json['tabletID'],
      deviceID: json['deviceID'],
      otoscopeID: json['otoscopeID'],
      tabletAppVersion: json['tabletAppVersion'],
      tabletAndroidVersion: json['tabletAndroidVersion'],
      centreId: json['centreId'],
      status:
          json['status'] != null
              ? statusFromApi(json['status'])
              : Status.inactive,
      createdAt:
          json['createdAt'] != null
              ? DateTime.parse(json['createdAt'])
              : DateTime.now(),
      updatedAt:
          json['updatedAt'] != null
              ? DateTime.parse(json['updatedAt'])
              : DateTime.now(),
      centre:
          json['centre'] != null ? CentreEntity.fromJson(json['centre']) : null,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'codeSequence': codeSequence,
      'tabletID': tabletID,
      'deviceID': deviceID,
      'otoscopeID': otoscopeID,
      'tabletAppVersion': tabletAppVersion,
      'tabletAndroidVersion': tabletAndroidVersion,
      'centreId': centreId,
      'status': status.name.toUpperCase(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'centre': centre?.toJson(),
    };
  }
}
